feat(tables): Rounded corner tables

aredridel opened the pull request:
Because rounding corners with a wrapping panel and `overflow: hidden` causes drop-down menus to not be able to cover a table, we have to implement rounded tables without clipping.

This moves the background onto the _cells_ of `th` and `td` within `thead` and `tfoot`, and creates a class for the header color style, orthogonal to the rounding.

The borders are applied to the corner cells themselves, since tables and thead/tbody/tfoot don't contain their children in the traditional layout sense.

aredridel committed:
feat(tables): Rounded corner table

aredridel committed:
fix(tables): Deprecate table-data

aredridel committed:
fix(tables): deprecate Alternate table

aredridel commented:
<img width="1000" alt="screen shot 2016-08-11 at 17 25 19" src="https://cloud.githubusercontent.com/assets/2876/17605889/9f31d32e-5fea-11e6-89dc-a71dee7c80e9.png">

stubbornella commented on src/pivotal-ui/components/tables/tables.scss:
```diff
@@ -73,12 +73,129 @@ table, .table {
+
+.table-rounded {
+  border-collapse: separate;
+  >thead:first-child>tr:first-child, >tfoot:first-child>tr:first-child, >tbody:first-child>tr:first-child, >tr:first-child {
```
please add line returns after each comma, it will make the individual selectors easier to parse. :)

stubbornella replied:
`>tfoot:first-child>tr:first-child` doesn't seem like a case that would happen IRL. Do we have designs that call for a table with only a footer and no `thead` or `tbody`?

aredridel commented:
@stubbornella any conceptual problems you see?

stubbornella commented:
Raquel and I will look more closely this morning! :)

stubbornella commented on src/pivotal-ui/components/tables/tables.scss:
```diff
@@ -73,12 +73,129 @@ table, .table {
+
+*/
+
+.table-header-gray-10 {
```
colors in classnames leads to 😭 in my experience

aredridel replied:
Could generate the whole set -- all the type colors & backgrounds have those as class names, etc.

jefflembeck replied:
I think, due to expected behavior w/ bootstrap, we should likely make these the defaults, then generate all other possibilities.

aredridel replied:
Alrighty.

stubbornella commented on src/pivotal-ui/components/tables/tables.scss:
```diff
@@ -73,12 +73,129 @@ table, .table {
+
+*/
+
+.table-rounded {
```
I wouldn't call this `rounded`, designers love to change the design. Maybe try something abstract that relates to the relationship between this table and another table. e.g. is it more or less important than a basic table? Simple? Basic? Highlight? Alternate? All words that define the relationship without referring to the way it looks.

aredridel replied:
It kinda _is_ the replacement for the replacement for the replacement for the basic table now. We're running out of low-meaning relative names.

aredridel replied:
`table-friendly` ?

stubbornella commented on src/pivotal-ui/components/tables/tables.scss:
```diff
@@ -73,12 +73,129 @@ table, .table {
+  border-collapse: separate;
+  >thead:first-child>tr:first-child, >tfoot:first-child>tr:first-child, >tbody:first-child>tr:first-child, >tr:first-child {
+    >th:first-child, >td:first-child {
+      border-top-left-radius: 10px !important;
```
I'd tend to simplify this a bit. It does a good job of covering all of the edge cases in the spec, but many of them feel pretty unlikely given our designs (and that we aren't building a framework for others to add to).

```
.table-alternate {
  thead,
  tbody:first-child,
  >tr:first-child {
    th:first-child,
    td:first-child {

    }
  }
}
```

It eliminates the exceptions for nested tables and also the tfoot being the first element in a table.

stubbornella commented on src/pivotal-ui/components/tables/tables.scss:
```diff
@@ -73,12 +73,129 @@ table, .table {
+      border-top: 1px solid $gray-8;
+    }
+  }
+  >thead>tr, >tbody>tr, >tfoot>tr, >tr {
```
If you didn't worry about nesting tables, this line:

```
>thead>tr, >tbody>tr, >tfoot>tr, >tr {
```

Could be simplified to:

```
tr {
```

aredridel replied:
Unfortunately, bootstrap. 😢

stubbornella commented on src/pivotal-ui/components/tables/tables.scss:
```diff
@@ -73,12 +73,129 @@ table, .table {
+      border-right: 1px solid $gray-8;
+    }
+  }
+  >thead:last-child>tr:last-child, >tfoot:last-child>tr:last-child, >tbody:last-child>tr:last-child, >tr:last-child {
```
I wouldn't cover the case where the `thead` is the last element in the table... seems like the edgiest of edge cases to me. :)

stubbornella replied:
how would you feel about...

```
.table-class {
  >:last-child {
    tr:last-child {
      bottom right rounding
    }
  }
}
```

aredridel replied:
Hm. That won't work in the unwrapped case, but I can simplify a bit because of that!

aredridel replied:
(AFAIK, IE still doesn't insert the `tbody` in the DOM)

aredridel committed:
Rename to table-friendly

aredridel commented:
Well fun. Simplifying in a couple of the suggested ways doesn't work because the default bootstrap `table` class has equally high precedence rules. 😢

aredridel committed:
Simplify the selectors a bit

aredridel committed:
Integrate the header colors in an easy-to-override way

We have no immediate use for other colors, and generating a full set
would bloat our CSS for a feature we aren't using.

We can however cross that bridge easily when we come to it.

jefflembeck commented:
🐑
